### PR TITLE
Add branch that reflects tutorial steps

### PR DESCRIPTION
**Pull request by jacek-foremski**
> [EZP-27292](https://jira.ez.no/browse/EZP-27292)
# 
This PR aims at creating a branch that has commits corresponding to each step from the tutorial.

### CODE REVIEW

File: Resources/misc/tag-steps.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	)
+declare -a tags=(
+	"create_the_bundle"
+	"structure_the_bundle"
+	"implement_the_tweet_value_class"
+	"implement_the_tweet_type_class"
+	"register_the_field_type_as_a_service"
+	"implement_the_legacy_storage_engine_converter"
+	"introduce_a_template"
+	)
```
Comment (mere-teresa) [inline Resources/misc/tag-steps.sh]: I would like to have tags with an enumeration.
Like this:
"step1_create_the_bundle"
"step2_structure_the_bundle"
"step3_implement_the_tweet_value_class"
"step4_implement_the_tweet_type_class"
"step5_register_the_field_type_as_a_service"
"step6_implement_the_legacy_storage_engine_converter"
"step7_introduce_a_template"

In the future, we could have additional step or modification of the order of the steps, it's the reason I don't want that the commit supports this enumeration, but the tags will support it.

File: Command/CreateTweetContentCommand.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
+<?php
+/**
+ * File containing the CreateTweetContentType class.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+
+namespace EzSystems\TweetFieldTypeBundle\Command;
```
Comment (bdunogier) [inline Command/CreateTweetContentCommand.php:9]: I'd move the command(s), as well as the XML payloads, to a separate commit (text suggestion: `CLI and REST helpers that create the content-type and content`), positioned after Implement the Legacy Storage Engine Converter. I suggest this position as they come usable at this point.

They should somehow come with explanations (maybe in the README) about how you use them (mostly for REST).

File: Command/CreateTweetContentTypeCommand.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
+<?php
+/**
+ * File containing the CreateTweetContentType class.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+
+namespace EzSystems\TweetFieldTypeBundle\Command;
```
Comment (bdunogier) [inline Command/CreateTweetContentTypeCommand.php:9]: Same comment than for `Command/CreateTweetContentCommand.php` (move to own commit).

File: Resources/misc/create-content-type.xml [hunk 4]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (bdunogier) [inline Resources/misc/create-content-type.xml:1]: Same comment than for the commands (and same commit in my opinion).

File: Resources/misc/tag-steps.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
```
Comment (bdunogier) [inline Resources/misc/tag-steps.sh:1]: I'd move this to its own commit. Not sure which is best though: first or last commit ? If it is first, changing any of it will require that we rebase all of the other commits. Maybe it is best if it is last ? It doesn't even need to be in the steps tags.

File: Tests/eZ/Publish/FieldType/TweetTest.php [hunk 6]
```diff
@@ -0,0 +1,199 @@
+<?php
```
Comment (bdunogier) [inline Tests/eZ/Publish/FieldType/TweetTest.php:1]: Must be moved to another commit. I was thinking about merging it with the code it tests, but @mere-teresa suggested a distinct commit, therefore located right after the addition of the tested code.

In addition, this should contain some kind of instructions about running those tests. PhpDoc in the test classes would work for me.

File: Tests/eZ/Publish/FieldType/TweetSPIIntegrationTest.php [hunk 7]
```diff
@@ -0,0 +1,144 @@
+<?php
```
Comment (bdunogier) [inline Tests/eZ/Publish/FieldType/TweetSPIIntegrationTest.php:1]: Same comment as for TweetTest below.

**Review by bdunogier — CHANGES_REQUESTED** [review]
The contents looks good, but the commits must be reorganized.

File: Command/CreateTweetContentCommand.php (hunk 2 above)
Comment (mere-teresa) [reply Command/CreateTweetContentCommand.php:9]: Yes, the README should contain some information about the non-guarantee for tags stability.
What will happen when you do, as a reader of the Tutorial a `git pull` and the tag has changed? It can be surprising.

File: Tests/eZ/Publish/FieldType/TweetTest.php (hunk 6 above)
Comment (mere-teresa) [reply Tests/eZ/Publish/FieldType/TweetTest.php:1]: Don't hesitate to use several commits for each tag, if needed. (If you work on the commits history and order, I put my foot in the door to add my wishes). I am thinking about something like this:

```
$> git log

commit <hash>
Creation of the class Type.php

commit <hash>
Adding this methods

commit <hash>
Tests for the class Type (optional in the Tutorial)
```


And the third commit is tagged as the stepN (_step4_implement_the_tweet_type_class_ ie) of the Tutorial. It put the question of the identification of the commits messages to have the tags on the table again.

Comment (jacek-foremski) [reply Tests/eZ/Publish/FieldType/TweetTest.php:1]: The potential problem I see with this approach would be that if you create a diff between tags, it would also show these optional files. But generally, I agree that tests should go right after commits that add tested functionality.

**Comment by jacek-foremski** [comment]
I made all the changes you requested, namely:
* added an enumeration to tags
* moved commands and XMLs to another commit
* added explanations in README about how to use commands and XMLs
* added warning to README about tags stability
* moved tags helper to the new last commit
* moved tests to new commits located directly after the code they are testing
* added instructions for running those Tests to the PhpDoc 

Please review the changes, especially the changes to the README.

**Comment by bdunogier** [comment]
I'd change `Introduce a template` to `Added field view ans field definition view templates`.

**Comment by jacek-foremski** [comment]
Changed commit message and tags helper accordingly.

File: Tests/eZ/Publish/FieldType/TweetTest.php (hunk 6 above)
Comment (bdunogier) [reply Tests/eZ/Publish/FieldType/TweetTest.php:1]: I see. It is a good point. But I wouldn't give too much weight to it, and would disregard it for now.

**Review by bdunogier — APPROVED** [review]
